burstphoto/align.swift

```diff
@@ -490,190 +1124,200 @@ func robust_merge(_ ref_texture: MTLTexture, _ ref_texture_blurred: MTLTexture,
+    command_encoder.setComputePipelineState(state)
+    let threads_per_grid = MTLSize(width: tile_info.n_tiles_x, height: tile_info.n_tiles_y, depth: 1)
+    let threads_per_thread_group = get_threads_per_thread_group(state, threads_per_grid)
+    command_encoder.setTexture(aligned_texture, index: 0)
+    command_encoder.setTexture(ref_texture, index: 1)
+    command_encoder.setTexture(rms_texture, index: 2)
+    command_encoder.setTexture(mismatch_texture, index: 3)
+    command_encoder.setBytes([Int32(tile_info.tile_size_merge)], length: MemoryLayout<Int32>.stride, index: 0)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    command_encoder.endEncoding()
+    command_buffer.commit()
     
-    // rescale output texture
-    let output_texture_uint16 = average_texture_sums(output_texture, n_images)
-    print("Time to align+merge images: ", Float(DispatchTime.now().uptimeNanoseconds - t) / 1_000_000_000)
+    return mismatch_texture
+}
+
+
+func normalize_mismatch(_ mismatch_texture: MTLTexture, _ mean_mismatch_buffer: MTLBuffer) {
+   
+    let command_buffer = command_queue.makeCommandBuffer()!
+    let command_encoder = command_buffer.makeComputeCommandEncoder()!
+    let state = normalize_mismatch_state
+    command_encoder.setComputePipelineState(state)
+    let threads_per_grid = MTLSize(width: mismatch_texture.width, height: mismatch_texture.height, depth: 1)
+    let threads_per_thread_group = get_threads_per_thread_group(state, threads_per_grid)
+    command_encoder.setTexture(mismatch_texture, index: 0)
+    command_encoder.setBuffer(mean_mismatch_buffer, offset: 0, index: 0)
+    command_encoder.dispatchThreads(threads_per_grid, threadsPerThreadgroup: threads_per_thread_group)
+    command_encoder.endEncoding()
+    command_buffer.commit()
+}
+
+
+func correct_hotpixels(_ textures: [MTLTexture]) {
     
-    // save the output image
-    try texture_to_dng(output_texture_uint16, in_url, out_url)
+    // generate simple average of all textures
+    let average_texture_descriptor = MTLTextureDescriptor.texture2DDescriptor(pixelFormat: .r32Float, width: textures[0].width, height: textures[0].height, mipmapped: false)
+    average_texture_descriptor.usage = [.shaderRead, .shaderWrite]
+    let average_texture = device.makeTexture(descriptor: average_texture_descriptor)!
+    fill_with_zeros(average_texture)
+              
+    // iterate over all images
+    for comp_idx in 0..<textures.count {
+            
+        add_texture(textures[comp_idx], average_texture, textures.count)
+    }
     
-    // delete the temporary dng directory
-    try FileManager.default.removeItem(atPath: tmp_dir)
+    // calculate mean value specific for each color channel
+    let mean_texture_buffer = texture_mean(convert_rgba(average_texture, 0, 0), "rgba")
     
-    return out_url
+    // iterate over all images and correct hot pixels in each texture
```

Comment:
This does not seem to be working properly.

This is one of the inputs with noise reduction off:
<img width="811" alt="input" src="https://user-images.githubusercontent.com/48962821/219974377-3f803965-bac0-4ae1-a43e-34fdba825fb9.png">

And this I the output:
<img width="811" alt="output" src="https://user-images.githubusercontent.com/48962821/219974383-ce92b28f-0121-484f-9e6f-5b1dec29059c.png">

Not only is the hot pixel not removed, but it's copy and pasted several times

Reply:
Based on our discussion, I'm marking this as done for now, and a potential fix for it will be added as part of #17.

Reply:
This issue has been resolved in the commit [Improved hot pixel correction](https://github.com/chris-rank/hdr-plus-swift/commit/7ee93ae216308050c1bfe5a48fd99ff25ff1e0a7) on the exposure branch.